File: Session4_Strings/palindromic-substrings.py
class Solution:
    def expandFromCtr(self, s, l, r):
        count = 0
        # Ensure no out of bounds condition
        while l >= 0 and r < len(s):
            # If the l and r pointer have same character
            if s[l] == s[r]:
                # Increase count due to presence of palindrome
                count += 1
                # Move left pointer outward (to the left)
                l -= 1
                # Move right pointer outward (to the right)
                r += 1
            # No more possible palindromes since char mismatch
            else:
                break
        return count
                
    
    def countSubstrings(self, s: str) -> int:
        totalCount = 0
        for i in range(len(s)):
            # Odd length palindrome case
            totalCount += self.expandFromCtr(s, i, i)
            # Even Length Palindrome case
            totalCount += self.expandFromCtr(s, i, i + 1)
        return totalCount

File: Session4_Strings/palindromic-substrings.java
class Solution {
    public int expandFromCtr(String s, int l, int r) {
        int count = 0;
        /* Ensure no out of bounds condition */
        while (l >= 0 && r < s.length()) {
            /* If the l and r pointer have same character */
            if (s.charAt(l) == s.charAt(r)) {
                /**
                 * Increase count due to presence of palindrome
                 * Move left pointer outward (to the left)
                 * Move right pointer outward (to the right)
                 */
                count++; l--; r++;
            }
            /* No more possible palindromes since char mismatch */
            else break;
        }   
        return count;
    }
    
    public int countSubstrings(String s) {
        int totalCount = 0;
        for (int i = 0; i < s.length(); i++) {
            /* Odd length palindrome case */
            totalCount += expandFromCtr(s, i, i);
            /* Even length palindrome case  */
            totalCount += expandFromCtr(s, i, i + 1);
        }
        return totalCount;
    }
}